### PR TITLE
Implement NumPy sorting routines (mSort).

### PR DESCRIPTION
**Pull request by us**
Implement `np.lexsort`, `np.msort`, `np.sort_complex`.
Related issue: #2079

### CODE REVIEW

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -2648,6 +2648,15 @@ def sort(a, axis=-1, kind='quicksort', order=None):
     return lax.sort(a, _canonicalize_axis(axis, ndim(a)))
 
 
+@_wraps(onp.lexsort)
+def lexsort(keys, axis=-1):
```
Comment (hawkinsp) [inline jax/numpy/lax_numpy.py]: Unfortunately the problem with this approach is that it won't work under jit transformation, and ideally we'd like that to be true in `jax.numpy` wherever possible. The reason it won't work under jit is that the control flow of sorted depends on the values of a, but they aren't known as concrete values at trace time.

I think, however that `lexsort` can be implemented using the underlying XLA sort operator.
https://www.tensorflow.org/xla/operation_semantics#sort
although the full generality of that operator is not exposed in JAX's `lax` library, which wraps it, so we'd first have to start by generalizing `lax.sort`. (I have a half-finished branch in which I did that myself.)

The same problem is true of `sort_complex` below; to implement `sort_complex`, you'd most likely need to use a tuple-sort of key-value pairs, which is something that XLA supports that is not plumbed into `lax` yet.

My main concern is that this project might make for a difficult starter project!

File: tests/lax_numpy_test.py [hunk 2]
```diff
@@ -1789,6 +1802,23 @@ def testArgsortManually(self):
     expected = onp.argsort(x)
     self.assertAllClose(expected, ans, check_dtypes=False)
 
+  def testMsortManually(self):
+    x = onp.array([16, 15, 23, 42, 8, 4])
+    ans = lnp.msort(x)
+    expected = onp.msort(x)
+    self.assertAllClose(expected, ans, check_dtypes=False)
```
Comment (hawkinsp) [inline tests/lax_numpy_test.py]: I'd suggest adding a call to `CompileAndCheck`, which verifies that a function produces the same result with and without `jit`. (If you try to do so for the other two functions, you will find that it fails, for the reasons I gave above. Try it!)

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Thanks for the PR!

I'd suggest submitting just `msort` in this PR; the others will need more work. I'm happy to walk you through what to do, but you might want to learn your way around JAX a bit first.

**Comment by us** [comment]
@hawkinsp Thank you for all advice! 
I will follow your suggestions.

**Note:** I have deleted the commit that include wrong sorting algorithms to avoid any confusion.

File: tests/lax_numpy_test.py (hunk 2 above)
Comment (us) [reply tests/lax_numpy_test.py]: I added `CompileAndCheck` and `CheckAgainstNumpy `.
I tried thank you!

File: jax/numpy/lax_numpy.py [hunk 3]
```diff
@@ -2664,6 +2673,17 @@ def argsort(a, axis=-1, kind='quicksort', order=None):
     return perm
 
 
+@_wraps(onp.msort)
+def msort(a):
```
[the remark this answers is not in the document]
Comment (us) [reply jax/numpy/lax_numpy.py]: Added.

**Review by hawkinsp — APPROVED** [review]
Thanks for the PR!

**Pull request merged**